File: Java/cp/MinimumWindowSubstring.java
/**
 * Minimum Window Substring 
 *          OR
 * Smallest Window in a String Containing All Characters of Another String
 * Given two strings string1 and string2, find the smallest substring in 
 * string1 containing all characters of string2. 
 */

import java.io.*;
public class MinimumWindowSubstring {
    
    static final int CHAR_COUNT = 26; 

    //Checking if all characters in pattern are contained in substring
    //or not
    public static boolean isValid(int[] str, int[] pattern) {
        for(int i=0; i<CHAR_COUNT; i++) {
            if(str[i] < pattern[i])
                return false; 
        }
        return true; 
    }

    public static String findSmallestSubstring(String str, String pattern) {

        //To store count of characters in str and pattern, respectively
        int strCount[] = new int[CHAR_COUNT];
        int ptrCount[] = new int[CHAR_COUNT]; 

        //Counting number of occurrences of each character in pattern
        for(int i=0; i<pattern.length(); i++) 
            ptrCount[pattern.charAt(i) - 'a']++; 

        int end, start;
        end = start = 0;
        String result = "-1"; //If no valid substring is found, "-1" is returned

        //start and end are pointers for start and end of substring
        while(end < str.length()) {

            //When substring does not contain all characters of pattern, we 
            //expand substring by adding characters from the end
            if(!isValid(strCount, ptrCount)) {
                strCount[str.charAt(end) - 'a']++; 
                end++; 
                continue; 
            } 

            //Updating substring contained in result
            if(result.equals("-1") || result.length() > (end - start))
                result = str.substring(start, end); 

            //To get length of smallest valid substring, we shorten the substring 
            //from front (i.e., eliminate characters from start)
            strCount[str.charAt(start) - 'a']--; 
            start++; 
        }

        //For the substrings which end at (str.length() - 1)
        while(isValid(strCount, ptrCount)) {
            if(result.equals("-1") || result.length() > (end - start))
                result = str.substring(start, end);
            strCount[str.charAt(start) - 'a']--; 
            start++;
        }

        return result; 
    }

    public static void main(String[] args) throws IOException {
        InputStreamReader read = new InputStreamReader(System.in);
	    BufferedReader buf = new BufferedReader(read); 

        //Taking input from user
        System.out.println("Enter a String text: ");
        String str = buf.readLine().trim();  
        System.out.println("Enter a String pattern: ");
        String pattern = buf.readLine().trim(); 

        //Printing output
        System.out.println("Smallest substring in text, containing all " +
                            "characters in pattern: ");
        System.out.println(findSmallestSubstring(str, pattern)); 
    }

}

/*
    Time complexity: O(N) (considering that isValid() method works in constant 
                        time as it makes a fixed number of comparisons) 
    Space complexity: O(CHAR_COUNT) 

    TEST CASES

    INPUT
    Enter a String text: 
    zaaskzaa
    Enter a String pattern: 
    zsk
    OUTPUT
    Smallest substring in text, containing all characters in pattern: 
    skz

    INPUT
    Enter a String text: 
    pompandpageantrygalore
    Enter a String pattern: 
    gtyn
    OUTPUT
    Smallest substring in text, containing all characters in pattern: 
    ntryg

 */

File: C-Plus-Plus/cp/minimum_window_substring.cpp
/**
 * Minimum Window Substring 
 *          OR
 * Smallest Window in a String Containing All Characters of Another String
 * Given two strings string1 and string2, find the smallest substring in 
 * string1 containing all characters of string2. 
 */

#include <bits/stdc++.h>
using namespace std;

string minWindow(string str, string pattern)
{ //unordered map for storing the characters in pattern that we need to check for in str
    unordered_map<char, int> letters;
    for (auto c : pattern)
        letters[c]++;
    //counts number of pattern'str letters in current window
    int count = 0;
    int low = 0, min_length = INT_MAX, min_start = 0;
    for (int high = 0; high < str.length(); high++) {
        if (letters[str[high]] > 0)
            count++; //means that this letter is in pattern
        //reduce the count for the letter on which we are currently
        letters[str[high]]--;
        if (count == pattern.length()) {
            //if current windows contains all of the letters in pattern
            while (low < high && letters[str[low]] < 0)
                letters[str[low]]++, low++;
            //move low ahead if its not of any significance
            if (min_length > high - low)
                //update the min length
                min_length = high - (min_start = low) + 1;
            //move low ahead and also increment the value
            letters[str[low++]]++;
            //count-- as we are moving low ahead & low pointed to a char in pattern before
            count--;
        }
    }
    //check for edge case & return the result
    return min_length == INT_MAX ? "" : str.substr(min_start, min_length);
}
int main()
{
    //taking input from user
    string str, pattern;
    cout << "Enter a String text: ";
    cin >> str;
    cout << "Enter a String pattern: ";
    cin >> pattern;
    cout << "Smallest substring in text, containing all characters in pattern: ";
    cout << minWindow(str, pattern);
}
/*
    Test Cases:

    INPUT:
    Enter a String text: saazksaa
    Enter a String pattern: szk
    OUTPUT:
    Smallest substring in text, containing all characters in pattern: zks

    INPUT:
    Enter a String text: letuscontributetoopensource
    Enter a String pattern: ubr
    OUTPUT:
    Smallest substring in text, containing all characters in pattern: ribu

    Time Complexity: O(N)
    Space Compleaxity: O(Charcter Count)

 */
